### ISSUE TITLE
Lambda Redshift Batch Load Failure NoSuchKey

### ISSUE THREAD

**Issue by stevesouders** [awslabs/aws-lambda-redshift-loader#73]
We're using this AWS Loader to copy logs into Redshift. We're writing ~10 logs per minute. They're constant but not huge; ~10 lines per log. Everything works great most of the time. Awesome!

But ~5 times per day we get a AWS SNS "Lambda Redshift Batch Load Failure" email. Some key values from the JSON in the email:

```
key: "2016-02-16T15:39:00.00-J0G3h_9mDcMYaB4AAAAA.log"
status: "error"
error:
    code: "XX000"
    file: "/home/awsrsqa/padb/src/sys/xen_execute.cpp"
    length: 810
    line: "6055"
    name: "error"
    routine: "pg_throw"
    severity: "ERROR"
    detail: [long string]
```

The error detail long string looks like this:

```
  -----------------------------------------------
  error:  S3ServiceException:The specified key does not exist.,Status 404,Error NoSuchKey,Rid snip,ExtRid snip,CanRetry 1
  code:      8001
  context:   S3 key being read : s3://foo-us-east.speedcurve.com/2016-02-16T15:39:00.000-Y4WGtElgPAZIkLkAAAAA.log
  query:     143771
  location:  table_s3_scanner.cpp:354
  process:   query0_30 [pid=17220]
  ----------------------------------------------- 
```

There are two log files mentioned in the email: in the "key" and "context" fields. None of the lines from those log files exist in our Redshift table. If I copy those logs to S3 manually they are imported without errors. 

Therefore, I believe our log format and Lambda setup are fine, and instead this might be a race condition of some sort. I read up on S3 consistency issues and wonder if that's what's going on here - the log file is copied up but the changes aren't fully replicated and so the attempt to find the file fails. 

Our goal is to automate a process to fix or retry these failures. Have you seen this before? Any ideas on a solution, or steps to debug further?

**Comment by IanMeyers** [comment]
Thanks for your mail. When you say that you have 10 files per minute of 10 lines each, does that mean that you are loading every 1 minute/10 files? Could you try an experiment and add `REGION 'us-east-1'` to your copy options in the configuration? I'm also querying with the Redshift team if there is a way to set the S3 endpoint to `s3-external-1` to force IAD endpoints...

**Comment by stevesouders** [comment]
Hi, Ian. Yes - we are loading ~10 files per minute. I get errors when adding `REGION 'us-east-1'` to our Copy Options. I tried adding it both in Dynamo DB and by re-running the NPM setup script. Perhaps I don't understand the format. Here's what I entered during the NPM setup script:

```
Additional Copy Options to be added > truncatecolumns ACCEPTINVCHARS REGION 'us-east-1'
```

Do they need to be comma separated or something? What would be the impact of this change? Our S3 bucket and Redshift are already both in us-east-1. Does this somehow mitigate possible race conditions?

**Comment by IanMeyers** [comment]
OK – that was just an experiment. I'll come back to you.

**Comment by stevesouders** [comment]
Hi, Ian.

Any update? We're still getting ~5 of these errors per day. 

Would altering these options have an effect?

```
How many files should be buffered before loading? > 10
How old should we allow a Batch to be before loading (seconds)? > 120
```

I notice that when the failure happens the data from 10 files is NOT written to Redshift, so reducing that would reduce the amount of lost data. Or perhaps we should drop that to "0" and instead just rely on the "how old" option to only process files after 120 seconds (when the consistency issues are resolved)?

Thanks, Steve

**Comment by IanMeyers** [comment]
Looks like my email reply didn't get attached. 

The issue should not be due to eventual consistency in S3, as the loader uses manifest files to indicate which files to load, and this takes care of ensuring files are resolved correctly. I'm wondering if the wildcard matching could be interfering. Do the files have any special characters like = in them, and/or are you using the hierarchical loading facility?

Please feel free to email on meyersi@amazon.com and we can maybe take a closer look.

**Comment by IanMeyers** [comment]
Closing as the issue is resolved with 2.4.0 and confirmed with the customer.
